aspnet/fundamentals/dependency-injection.rst

```diff
@@ -1,10 +1,291 @@
+
+With regard to data access specifically, you can easily inject Entity Framework ``DbContext`` types into your controllers, assuming you've configured EF in your ``Startup`` class. However, I recommend avoiding direct dependency on ``DbContext`` in your application. Instead, depend on an abstraction (like a Repository interface), and restrict knowledge of EF (or any other specific data access technology) to the implementation of this interface. This will reduce the coupling between your application and a particular data access strategy, and will make :doc:`testing` your application code much easier.
+
+Replacing the default services container
+----------------------------------------
+
+The built-in services container is meant to serve the basic needs of the framework and most consumer applications built on it. However, developers who wish to replace the built-in container with their preferred container can easily do so. The ``ConfigureServices`` method typically returns ``void``, but if its signature is changed to return ``IServiceProvider``, a different container can be configured and returned. There are `many IOC containers available for .NET <https://github.com/danielpalme/IocPerformance/blob/master/README.md>`_. This article will attempt to add links to DNX implementations of containers as they become available. In this example, the `StructureMap.Dnx <https://github.com/structuremap/structuremap.dnx>`_ package is used.
+
+First, add the appropriate container to the dependencies property in project.json:
+
+.. code-block:: javascript
+
+	"dependencies" : {
+		"StructureMap.Dnx": "0.4.0-alpha"
+	},
+
+.. note:: Currently StructureMap does not support .NET Core.
+
+Next, configure the container in ``ConfigureServices`` and return an ``IServiceProvider``:
+
+.. code-block:: c#
+
+	public IServiceProvider ConfigureServices(IServiceCollection services)
+	{
+		services.AddMvc();
+		// add other services
+		
+		// Add StructureMap
+		var container = new StructureMap.Container();
+		container.Populate(services);
+		return container.GetInstance<IServiceProvider>();
+	}
+
+.. note:: If you add StructureMap to the sample project, you will need to add a ``[DefaultConstructor]`` attribute to the default constructor for ``Operation``.
+
+.. list-table:: ASP.NET 5 / DNX Containers (in alphabetical order)
+   :header-rows: 1
+
+   * - Package (Nuget)
+     - Setup Instructions
+   * - `Autofac.Dnx <http://www.nuget.org/packages/Autofac.Dnx/>`_
+     - `Autofac.org <http://autofac.org/>`_
+   * - `StructureMap.Dnx <http://www.nuget.org/packages/StructureMap.Dnx>`_
+     - `Getting Started <https://github.com/structuremap/structuremap.dnx/blob/master/README.md>`_
+
+Recommendations
+---------------
+
+When working with dependency injection, keep the following recommendations in mind:
+
+- DI is for objects that have complex dependencies. Controllers, services, adapters, and repositories are all examples of objects that might be added to DI.
```

Comment:
More like different uses of the term. It's not unusual to have services within an application (e.g. http://gorodinski.com/blog/2012/04/14/services-in-domain-driven-design-ddd/, https://lostechies.com/jimmybogard/2008/08/21/services-in-domain-driven-design/). I could call it "Application Services" or "Domain Services" though if you think it's at all confusing as-is.